File: util/android/src/main/java/com/dmi/util/setting/AbstractSettingsApplier.java
package com.dmi.util.setting;

import java.util.ArrayList;
import java.util.List;

public abstract class AbstractSettingsApplier {
    private final List<AbstractSettings.Setting<?>> settings = new ArrayList<>();

    public void startListen() {
        listen();
    }

    public void stopListen() {
        for (AbstractSettings.Setting<?> setting : settings) {
            setting.setListener(null);
        }
    }

    protected abstract void listen();

    protected <T> void listen(AbstractSettings.Setting<T> setting, SettingListener<T> listener) {
        setting.setListener(listener);
        settings.add(setting);
    }
}

File: util/util-android/src/main/kotlin/com/dmi/util/setting/AbstractSettingsApplier.kt
package com.dmi.util.setting

import java.util.*

abstract class AbstractSettingsApplier {
    private val settings = ArrayList<AbstractSettings.Setting<*>>()

    fun startListen() {
        listen()
    }

    fun stopListen() {
        for (setting in settings) {
            setting.removeListener()
        }
    }

    protected abstract fun listen()

    protected fun <T: Any> listen(setting: AbstractSettings.Setting<T>, listener: SettingListener<T>) {
        setting.setListener(listener)
        settings.add(setting)
    }
}